requests/models.py

```diff
@@ -365,11 +365,19 @@ def prepare_url(self, url, params):
         if not host:
             raise InvalidURL("Invalid URL %r: No host supplied" % url)
 
-        # Only want to apply IDNA to the hostname
-        try:
-            host = idna.encode(host, uts46=True).decode('utf-8')
-        except (UnicodeError, idna.IDNAError):
-            raise InvalidURL('URL has an invalid label.')
+        # In general, we want to IDNA encode everything, as that allows users
+        # to automatically get correct behaviour in the majority of cases.
+        # However, we’re quite strict about IDNA encoding, so if a user has
+        # already IDNA encoded their domain name, we shouldn’t bother validating
+        # their work. We determine something *looks* IDNA encoded by checking
+        # for the IDNA-encoding tag (xn--) at the start of the hostname, and
+        # confirming that the hostname is entirely ASCII characters.
+        if not host.startswith('xn--') or not is_ascii(host):
```

Comment:
@Lukasa, sorry I deleted this the first time, thought I'd made a mistake but I've arrived back at the same conclusion. I'm onboard with `u'xn--'`. 

While `unicode_is_ascii` is definitely doable/simpler, I do have a couple outstanding questions:

1. The new function will only be the single try/except block, so does it really warrant being a util function at that point?

2. While we are now explicitly stating that this is only for unicode, if you supply the function a bytestring without the current checks, it will evaluate to `True` in Python 2 and `False` in Python 3. If this ever gets reused in a place were people are *pretty* sure it will always be unicode but may not, that's a tricky place for errors to crop up.

The repro is simply `b'test'.encode('ascii')`. Python 2 will let you do this all day, Python 3 raises an `AttributeError`. We could alternatively wrap the try/except in a conditional block to check if it's a string, but that has the inverse problem of string in the format `'my string'` being `False` in Python 2 and `True` in Python 3.